--- a/net/minecraft/block/state/pattern/BlockPattern.java
+++ b/net/minecraft/block/state/pattern/BlockPattern.java
@@ -86,9 +86,9 @@
 
         for (BlockPos blockpos : BlockPos.getAllInBox(pos, pos.add(i - 1, i - 1, i - 1)))
         {
-            for (EnumFacing enumfacing : EnumFacing.values())
+            for (EnumFacing enumfacing : EnumFacing.ENUM_FACING_VALUES)
             {
-                for (EnumFacing enumfacing1 : EnumFacing.values())
+                for (EnumFacing enumfacing1 : EnumFacing.ENUM_FACING_VALUES)
                 {
                     if (enumfacing1 != enumfacing && enumfacing1 != enumfacing.getOpposite())
                     {
