--- a/net/minecraft/command/impl/EnchantCommand.java
+++ b/net/minecraft/command/impl/EnchantCommand.java
@@ -1,5 +1,6 @@
 package net.minecraft.command.impl;
 
+import carpet.settings.CarpetSettings;
 import com.mojang.brigadier.CommandDispatcher;
 import com.mojang.brigadier.arguments.IntegerArgumentType;
 import com.mojang.brigadier.exceptions.CommandSyntaxException;
@@ -54,7 +55,7 @@
 
     private static int enchant(CommandSource source, Collection <? extends Entity > targets, Enchantment enchantmentIn, int level) throws CommandSyntaxException
     {
-        if (level > enchantmentIn.getMaxLevel())
+        if (!CarpetSettings.enchantCommandNoRestriction && level > enchantmentIn.getMaxLevel())  // TISCM added rule check
         {
             throw INVALID_LEVEL.create(level, enchantmentIn.getMaxLevel());
         }
@@ -71,7 +72,7 @@
 
                     if (!itemstack.isEmpty())
                     {
-                        if (enchantmentIn.canApply(itemstack) && EnchantmentHelper.areAllCompatibleWith(EnchantmentHelper.getEnchantments(itemstack).keySet(), enchantmentIn))
+                        if (CarpetSettings.enchantCommandNoRestriction || (enchantmentIn.canApply(itemstack) && EnchantmentHelper.areAllCompatibleWith(EnchantmentHelper.getEnchantments(itemstack).keySet(), enchantmentIn)))  // TISCM added rule check
                         {
                             itemstack.addEnchantment(enchantmentIn, level);
                             ++i;
