--- a/net/minecraft/entity/projectile/EntityFishHook.java
+++ b/net/minecraft/entity/projectile/EntityFishHook.java
@@ -2,6 +2,10 @@
 
 import java.util.Collections;
 import java.util.List;
+
+import carpet.patches.EntityPlayerMPFake;
+import gca.GcaSetting;
+import gca.tools.FakePlayerAutoFish;
 import net.minecraft.advancements.CriteriaTriggers;
 import net.minecraft.block.Block;
 import net.minecraft.entity.Entity;
@@ -448,6 +452,10 @@
                 worldserver.spawnParticle(Particles.BUBBLE, this.posX, d3, this.posZ, (int)(1.0F + this.width * 20.0F), (double)this.width, 0.0D, (double)this.width, (double)0.2F);
                 worldserver.spawnParticle(Particles.FISHING, this.posX, d3, this.posZ, (int)(1.0F + this.width * 20.0F), (double)this.width, 0.0D, (double)this.width, (double)0.2F);
                 this.ticksCatchable = MathHelper.nextInt(this.rand, 20, 40);
+                Entity entity = this.getAngler();
+                if (GcaSetting.fakePlayerAutoFish && entity instanceof EntityPlayerMPFake) {
+                    FakePlayerAutoFish.autoFish((EntityPlayer) entity);
+                }
             }
         }
         else if (this.ticksCaughtDelay > 0)
