--- a/net/minecraft/entity/projectile/EntityFishHook.java
+++ b/net/minecraft/entity/projectile/EntityFishHook.java
@@ -425,7 +425,7 @@
                 double d0 = this.posX + (double)(f1 * (float)this.ticksCatchableDelay * 0.1F);
                 double d1 = (double)((float)MathHelper.floor(this.getBoundingBox().minY) + 1.0F);
                 double d2 = this.posZ + (double)(f2 * (float)this.ticksCatchableDelay * 0.1F);
-                Block block = worldserver.getBlockState(new BlockPos(d0, d1 - 1.0D, d2)).getBlock();
+                Block block = this.getBlockState(new BlockPos(d0, d1 - 1.0D, d2)).getBlock();
 
                 if (block == Blocks.WATER)
                 {
@@ -475,7 +475,7 @@
                 double d4 = this.posX + (double)(MathHelper.sin(f6) * f7 * 0.1F);
                 double d5 = (double)((float)MathHelper.floor(this.getBoundingBox().minY) + 1.0F);
                 double d6 = this.posZ + (double)(MathHelper.cos(f6) * f7 * 0.1F);
-                Block block1 = worldserver.getBlockState(new BlockPos((int)d4, (int)d5 - 1, (int)d6)).getBlock();
+                Block block1 = this.getBlockState(new BlockPos((int)d4, (int)d5 - 1, (int)d6)).getBlock();
 
                 if (block1 == Blocks.WATER)
                 {
