--- a/net/minecraft/nbt/NBTTagIntArray.java
+++ b/net/minecraft/nbt/NBTTagIntArray.java
@@ -5,6 +5,9 @@
 import java.io.IOException;
 import java.util.Arrays;
 import java.util.List;
+
+import me.jellysquid.mods.lithium.common.nbt.io.NbtFastReader;
+import me.jellysquid.mods.lithium.common.nbt.io.NbtFastWriter;
 import net.minecraft.util.text.ITextComponent;
 import net.minecraft.util.text.TextComponentString;
 import org.apache.commons.lang3.ArrayUtils;
@@ -152,4 +155,28 @@
     {
         this.intArray = ArrayUtils.remove(this.intArray, p_197649_1_);
     }
+
+    @Override
+    public void serialize(NbtFastWriter writer) {
+        writer.writeInt(this.intArray.length);
+
+        for (int i : this.intArray) {
+            writer.writeInt(i);
+        }
+    }
+
+    @Override
+    public void deserialize(NbtFastReader reader, int level, NBTSizeTracker positionTracker) {
+        positionTracker.read(192L);
+
+        int size = reader.readInt();
+
+        positionTracker.read(32 * size);
+
+        this.intArray = new int[size];
+
+        for(int i = 0; i < size; ++i) {
+            this.intArray[i] = reader.readInt();
+        }
+    }
 }
