--- a/net/minecraft/nbt/NBTTagByte.java
+++ b/net/minecraft/nbt/NBTTagByte.java
@@ -3,6 +3,9 @@
 import java.io.DataInput;
 import java.io.DataOutput;
 import java.io.IOException;
+
+import me.jellysquid.mods.lithium.common.nbt.io.NbtFastReader;
+import me.jellysquid.mods.lithium.common.nbt.io.NbtFastWriter;
 import net.minecraft.util.text.ITextComponent;
 import net.minecraft.util.text.TextComponentString;
 
@@ -68,6 +71,17 @@
         return (new TextComponentString(String.valueOf((int)this.data))).appendSibling(itextcomponent).applyTextStyle(SYNTAX_HIGHLIGHTING_NUMBER);
     }
 
+    @Override
+    public void serialize(NbtFastWriter writer) {
+        writer.writeByte(this.data);
+    }
+
+    @Override
+    public void deserialize(NbtFastReader reader, int level, NBTSizeTracker sizeTracker) {
+        sizeTracker.read(72L);
+        this.data =reader.readByte();
+    }
+
     public long getLong()
     {
         return (long)this.data;
