--- a/net/minecraft/tileentity/TileEntityEndGateway.java
+++ b/net/minecraft/tileentity/TileEntityEndGateway.java
@@ -320,7 +320,7 @@
     {
         int i = 0;
 
-        for (EnumFacing enumfacing : EnumFacing.values())
+        for (EnumFacing enumfacing : (me.jellysquid.mods.lithium.LithiumConfig.ALLOC_ENUM_VALUES ? EnumFacing.ENUM_FACING_VALUES : EnumFacing.values()))
         {
             i += this.shouldRenderFace(enumfacing) ? 1 : 0;
         }
