--- a/net/minecraft/server/management/PlayerChunkMapEntry.java
+++ b/net/minecraft/server/management/PlayerChunkMapEntry.java
@@ -1,9 +1,15 @@
 package net.minecraft.server.management;
 
+import carpet.settings.CarpetSettings;
+import carpet.utils.LightTrackingHooks;
 import com.google.common.collect.Lists;
 import java.util.List;
+import java.util.Map;
+import java.util.Set;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
+
+import it.unimi.dsi.fastutil.shorts.ShortLinkedOpenHashSet;
 import net.minecraft.entity.player.EntityPlayerMP;
 import net.minecraft.network.Packet;
 import net.minecraft.network.play.server.SPacketBlockChange;
@@ -25,14 +31,26 @@
     private final PlayerChunkMap playerChunkMap;
     private final List<EntityPlayerMP> players = Lists.newArrayList();
     private final ChunkPos pos;
-    private final short[] changedBlocks = new short[64];
+    // TISCM chunkUpdatePacketThreshold removed final
+    private short[] changedBlocks = new short[64];
     @Nullable
     private Chunk chunk;
     private int changes;
     private int changedSectionFilter;
     private long lastUpdateInhabitedTime;
     private boolean sentToPlayers;
+    
+    // [TISCM] Newlight - start
+    private java.util.Map<EntityPlayerMP, int[]> lightTrackings = new java.util.HashMap<>();
+    private final java.util.Map<EntityPlayerMP, int[]> neighborLightTrackings = new java.util.HashMap<>();
+    // [TISCM] Newlight - end
 
+    // TISCM chunkUpdatePacketThreshold
+    private boolean ruleEnabled$CUPT;
+    private int ruleValue$CUPT;
+    private final Set<Short> changedBlocksSet$CUPT = new ShortLinkedOpenHashSet();
+    // TISCM chunkUpdatePacketThreshold ends
+
     public PlayerChunkMapEntry(PlayerChunkMap mapIn, int chunkX, int chunkZ)
     {
         this.playerChunkMap = mapIn;
@@ -40,8 +58,29 @@
         ChunkProviderServer chunkproviderserver = mapIn.getWorld().getChunkProvider();
         chunkproviderserver.touch(chunkX, chunkZ);
         this.chunk = chunkproviderserver.getChunk(chunkX, chunkZ, true, false);
+
+        // TISCM chunkUpdatePacketThreshold
+        this.updateRuleStatus$CUPT();
     }
 
+    private void updateRuleStatus$CUPT()
+    {
+        this.ruleValue$CUPT = CarpetSettings.chunkUpdatePacketThreshold;
+        this.ruleEnabled$CUPT = this.ruleValue$CUPT != CarpetSettings.VANILLA_CHUNK_UPDATE_PACKET_THRESHOLD;
+        // rollback to vanilla array
+        if (!this.ruleEnabled$CUPT)
+        {
+            this.changedBlocks = new short[CarpetSettings.VANILLA_CHUNK_UPDATE_PACKET_THRESHOLD];
+        }
+    }
+
+    // TISCM chunkUpdatePacketThreshold
+    private int getChunkPacketThreshold()
+    {
+        // 64 is the vanilla value
+        return this.ruleEnabled$CUPT ? this.ruleValue$CUPT : 64;
+    }
+
     public ChunkPos getPos()
     {
         return this.pos;
@@ -64,6 +103,9 @@
 
             if (this.sentToPlayers)
             {
+                // [TISCM] Newlight
+                if (CarpetSettings.newLight)
+                    LightTrackingHooks.addPlayer(player, this, this.playerChunkMap);
                 this.sendToPlayer(player);
             }
         }
@@ -75,6 +117,9 @@
         {
             if (this.sentToPlayers)
             {
+                // [TISCM] Newlight
+                if (CarpetSettings.newLight)
+                    LightTrackingHooks.removePlayer(player, this, this.playerChunkMap);
                 player.connection.sendPacket(new SPacketUnloadChunk(this.pos.x, this.pos.z));
             }
 
@@ -126,8 +171,16 @@
 
                 for (EntityPlayerMP entityplayermp : this.players)
                 {
+                    // [TISCM] Newlight
+                    if (CarpetSettings.newLight)
+                        LightTrackingHooks.addPlayer(entityplayermp, this, this.playerChunkMap);
+                    
                     entityplayermp.connection.sendPacket(packet);
                     this.playerChunkMap.getWorld().getEntityTracker().sendLeashedEntitiesInChunk(entityplayermp, this.chunk);
+    
+                    // [TISCM] Newlight
+                    if (CarpetSettings.newLight)
+                        LightTrackingHooks.sendLightTrackings(this, entityplayermp, this.playerChunkMap);
                 }
             }
 
@@ -141,6 +194,10 @@
         {
             player.connection.sendPacket(new SPacketChunkData(this.chunk, 65535));
             this.playerChunkMap.getWorld().getEntityTracker().sendLeashedEntitiesInChunk(player, this.chunk);
+    
+            // [TISCM] Newlight
+            if (CarpetSettings.newLight)
+                LightTrackingHooks.sendLightTrackings(this, player, this.playerChunkMap);
         }
     }
 
@@ -167,19 +224,31 @@
 
             this.changedSectionFilter |= 1 << (y >> 4);
 
-            if (this.changes < 64)
+            // TISCM changedBlocksList
+            // modified constant 64 into CarpetSettings.chunkPacketThreshold
+            if (this.changes < this.getChunkPacketThreshold())
             {
                 short short1 = (short)(x << 12 | z << 8 | y);
 
-                for (int i = 0; i < this.changes; ++i)
+                // TISCM changedBlocksList
+                if (this.ruleEnabled$CUPT)
                 {
-                    if (this.changedBlocks[i] == short1)
+                    this.changedBlocksSet$CUPT.add(short1);
+                    this.changes = this.changedBlocksSet$CUPT.size();
+                }
+                // vanilla
+                else
+                {
+                    for (int i = 0; i < this.changes; ++i)
                     {
-                        return;
+                        if (this.changedBlocks[i] == short1)
+                        {
+                            return;
+                        }
                     }
-                }
 
-                this.changedBlocks[this.changes++] = short1;
+                    this.changedBlocks[this.changes++] = short1;
+                }
             }
         }
     }
@@ -201,6 +270,18 @@
         {
             if (this.changes != 0)
             {
+                // TISCM chunkUpdatePacketThreshold
+                if (this.ruleEnabled$CUPT)
+                {
+                    this.changedBlocks = new short[this.changedBlocksSet$CUPT.size()];
+                    int i = 0;
+                    for (Short s: this.changedBlocksSet$CUPT)
+                    {
+                        this.changedBlocks[i++] = s;
+                    }
+                    this.changedBlocksSet$CUPT.clear();
+                }
+
                 if (this.changes == 1)
                 {
                     int i = (this.changedBlocks[0] >> 12 & 15) + this.pos.x * 16;
@@ -214,9 +295,14 @@
                         this.sendBlockEntity(this.playerChunkMap.getWorld().getTileEntity(blockpos));
                     }
                 }
-                else if (this.changes == 64)
+                // TISCM chunkUpdatePacketThreshold
+                // modified constant 64
+                else if (this.changes == this.getChunkPacketThreshold())
                 {
                     this.sendPacket(new SPacketChunkData(this.chunk, this.changedSectionFilter));
+                    // [TISCM] Newlight
+                    if (CarpetSettings.newLight)
+                        LightTrackingHooks.sendLightTrackings(this, this.changedSectionFilter, this.playerChunkMap);
                 }
                 else
                 {
@@ -239,6 +325,9 @@
                 this.changes = 0;
                 this.changedSectionFilter = 0;
             }
+
+            // TISCM chunkUpdatePacketThreshold
+            this.updateRuleStatus$CUPT();
         }
     }
 
@@ -309,4 +398,15 @@
 
         return d0;
     }
+    
+    // [TISCM] Newlight - getters
+    public Map<EntityPlayerMP, int[]> getLightTrackings()
+    {
+        return lightTrackings;
+    }
+    
+    public Map<EntityPlayerMP, int[]> getNeighborLightTrackings()
+    {
+        return neighborLightTrackings;
+    }
 }
