--- a/net/minecraft/server/management/PlayerInteractionManager.java
+++ b/net/minecraft/server/management/PlayerInteractionManager.java
@@ -1,5 +1,8 @@
 package net.minecraft.server.management;
 
+import carpet.CarpetServer;
+import carpet.helpers.BlockRotator;
+import carpet.worldedit.CarpetWorldEditAccess;
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockChest;
 import net.minecraft.block.BlockCommandBlock;
@@ -241,6 +244,12 @@
         {
             iblockstate.getBlock().onPlayerDestroy(this.world, pos, iblockstate);
         }
+        //CM
+        if(flag)
+        {
+            CarpetServer.scriptServer.events.onBlockBroken(player, world, pos, iblockstate);
+        }
+        // end CM
 
         return flag;
     }
@@ -312,6 +321,14 @@
 
     public EnumActionResult processRightClick(EntityPlayer player, World worldIn, ItemStack stack, EnumHand hand)
     {
+        // CM worldedit
+        ActionResult<ItemStack> weResult = CarpetWorldEditAccess.onRightClickAir(player, worldIn, hand);
+        if (weResult != null && weResult.getType() != EnumActionResult.PASS)
+        {
+            return weResult.getType();
+        }
+        // CM worldedit ends
+
         if (this.gameType == GameType.SPECTATOR)
         {
             return EnumActionResult.PASS;
@@ -366,6 +383,14 @@
 
     public EnumActionResult processRightClickBlock(EntityPlayer player, World worldIn, ItemStack stack, EnumHand hand, BlockPos pos, EnumFacing facing, float hitX, float hitY, float hitZ)
     {
+        // CM worldedit
+        EnumActionResult weResult = CarpetWorldEditAccess.onRightClickBlock(player, worldIn, hand, facing, pos);
+        if (weResult != EnumActionResult.PASS)
+        {
+            return weResult;
+        }
+        // CM worldedit ends
+
         IBlockState iblockstate = worldIn.getBlockState(pos);
 
         if (this.gameType == GameType.SPECTATOR)
@@ -401,11 +426,20 @@
             boolean flag = !player.getHeldItemMainhand().isEmpty() || !player.getHeldItemOffhand().isEmpty();
             boolean flag1 = player.isSneaking() && flag;
 
-            if (!flag1 && iblockstate.onBlockActivated(worldIn, pos, player, hand, facing, hitX, hitY, hitZ))
+            if (!flag1)
             {
-                return EnumActionResult.SUCCESS;
+                //flip method will check for flippinCactus setting
+                Boolean flipped = BlockRotator.flipBlockWithCactus(worldIn, pos, iblockstate, player, hand, facing, hitX, hitY, hitZ);
+                if (flipped)
+                {
+                    return EnumActionResult.PASS;
+                }
+                if (iblockstate.onBlockActivated(worldIn, pos, player, hand, facing, hitX, hitY, hitZ))
+                {
+                    return EnumActionResult.SUCCESS;
+                }
             }
-            else if (!stack.isEmpty() && !player.getCooldownTracker().hasCooldown(stack.getItem()))
+            if (!stack.isEmpty() && !player.getCooldownTracker().hasCooldown(stack.getItem()))
             {
                 ItemUseContext itemusecontext = new ItemUseContext(player, player.getHeldItem(hand), pos, facing, hitX, hitY, hitZ);
 
