--- a/net/minecraft/pathfinding/WalkAndSwimNodeProcessor.java
+++ b/net/minecraft/pathfinding/WalkAndSwimNodeProcessor.java
@@ -294,7 +294,7 @@
 
         if (pathnodetype == PathNodeType.WATER)
         {
-            for (EnumFacing enumfacing : EnumFacing.values())
+            for (EnumFacing enumfacing : (me.jellysquid.mods.lithium.LithiumConfig.ALLOC_ENUM_VALUES ? EnumFacing.ENUM_FACING_VALUES : EnumFacing.values()))
             {
                 PathNodeType pathnodetype2 = this.getPathNodeTypeRaw(blockaccessIn, x + enumfacing.getXOffset(), y + enumfacing.getYOffset(), z + enumfacing.getZOffset());
 
