--- a/net/minecraft/world/WorldServer.java
+++ b/net/minecraft/world/WorldServer.java
@@ -1,5 +1,9 @@
 package net.minecraft.world;
 
+import carpet.helpers.TickSpeed;
+import carpet.settings.CarpetSettings;
+import carpet.utils.CarpetProfiler;
+import carpet.logging.logHelpers.TileTickListLogHelper;
 import com.google.common.collect.Lists;
 import com.google.common.collect.Maps;
 import com.google.common.util.concurrent.ListenableFuture;
@@ -14,6 +18,9 @@
 import java.util.stream.Stream;
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
+
+import me.jellysquid.mods.lithium.common.world.scheduler.LithiumServerTickScheduler;
+import me.jellysquid.mods.lithium.config.LiConfig;
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockEventData;
 import net.minecraft.block.state.IBlockState;
@@ -98,17 +105,13 @@
     private int updateEntityTick;
     private final Teleporter worldTeleporter;
     private final WorldEntitySpawner entitySpawner = new WorldEntitySpawner();
-    private final ServerTickList<Block> pendingBlockTicks = new ServerTickList<>(this, (p_205341_0_) ->
-    {
-        return p_205341_0_ == null || p_205341_0_.getDefaultState().isAir();
-    }, IRegistry.BLOCK::getKey, IRegistry.BLOCK::getOrDefault, this::tickBlock);
-    private final ServerTickList<Fluid> pendingFluidTicks = new ServerTickList<>(this, (p_205774_0_) ->
-    {
-        return p_205774_0_ == null || p_205774_0_ == Fluids.EMPTY;
-    }, IRegistry.FLUID::getKey, IRegistry.FLUID::getOrDefault, this::tickFluid);
+    private final ServerTickList<Block> pendingBlockTicks;
+    private final ServerTickList<Fluid> pendingFluidTicks;
     protected final VillageSiege villageSiege = new VillageSiege(this);
     ObjectLinkedOpenHashSet<BlockEventData> blockEventQueue = new ObjectLinkedOpenHashSet<>();
     private boolean insideTick;
+    /** Lithium: avoidAlloc */
+    private final BlockPos.MutableBlockPos mutableBlockPosForRandomTicks = new BlockPos.MutableBlockPos();
 
     public WorldServer(MinecraftServer server, ISaveHandler p_i49819_2_, WorldSavedDataStorage p_i49819_3_, WorldInfo p_i49819_4_, DimensionType p_i49819_5_, Profiler p_i49819_6_)
     {
@@ -122,6 +125,35 @@
         this.calculateInitialSkylight();
         this.calculateInitialWeather();
         this.getWorldBorder().setSize(server.getMaxWorldSize());
+        if (LiConfig.World.world_useLithiumTickScheduler)
+        {
+            this.pendingBlockTicks = new LithiumServerTickScheduler<>(this, (block) ->
+                    block == null || block.getDefaultState().isAir(),
+                    IRegistry.BLOCK::getKey,
+                    IRegistry.BLOCK::getOrDefault,
+                    this::tickBlock
+            );
+            this.pendingFluidTicks = new LithiumServerTickScheduler<>(this, (fluid) ->
+                    fluid == null || fluid == Fluids.EMPTY,
+                    IRegistry.FLUID::getKey,
+                    IRegistry.FLUID::getOrDefault,
+                    this::tickFluid
+            );
+        }
+        else {
+            this.pendingBlockTicks = new ServerTickList<>(this, (block) ->
+                    block == null || block.getDefaultState().isAir(),
+                    IRegistry.BLOCK::getKey,
+                    IRegistry.BLOCK::getOrDefault,
+                    this::tickBlock
+            );
+            this.pendingFluidTicks = new ServerTickList<>(this, (fluid) ->
+                    fluid == null || fluid == Fluids.EMPTY,
+                    IRegistry.FLUID::getKey,
+                    IRegistry.FLUID::getOrDefault,
+                    this::tickFluid
+            );
+        }
     }
 
     public WorldServer init()
@@ -191,13 +223,24 @@
             this.wakeAllPlayers();
         }
 
+        //CM profiler
+        int did = this.dimension.getType().getId();
+        String world_name = (did==0)?"Overworld":((did<0?"The Nether":"The End"));
+        // CM end
+
+        if (TickSpeed.process_entities)
+        { // [CM] extra indent to skip processing of entities
         this.profiler.startSection("spawner");
+        CarpetProfiler.start_section(world_name, "Spawning");
 
         if (this.getGameRules().getBoolean("doMobSpawning") && this.worldInfo.getGenerator() != WorldType.DEBUG_ALL_BLOCK_STATES)
         {
             this.entitySpawner.findChunksForSpawning(this, this.spawnHostileMobs, this.spawnPeacefulMobs, this.worldInfo.getGameTime() % 400L == 0L);
             this.getChunkProvider().spawnMobs(this, this.spawnHostileMobs, this.spawnPeacefulMobs);
         }
+        CarpetProfiler.end_current_section();
+        }
+        // [CM] end extra indent
 
         this.profiler.endStartSection("chunkSource");
         this.chunkProvider.tick(hasTimeLeft);
@@ -208,24 +251,42 @@
             this.setSkylightSubtracted(j);
         }
 
+        if (TickSpeed.process_entities)
+        { // CM extra indent to skip processing of entities
         this.worldInfo.setGameTime(this.worldInfo.getGameTime() + 1L);
-
         if (this.getGameRules().getBoolean("doDaylightCycle"))
         {
             this.worldInfo.setDayTime(this.worldInfo.getDayTime() + 1L);
         }
 
         this.profiler.endStartSection("tickPending");
+        CarpetProfiler.start_section(world_name, "Blocks");
         this.tickPending();
+        CarpetProfiler.end_current_section();
+        } // end extra indent
         this.profiler.endStartSection("tickBlocks");
+        CarpetProfiler.start_section(world_name, "Blocks");
         this.tickBlocks();
+        CarpetProfiler.end_current_section();
         this.profiler.endStartSection("chunkMap");
         this.playerChunkMap.tick();
+
+        if (TickSpeed.process_entities)
+        { // CM extra indent to skip processing of entities
         this.profiler.endStartSection("village");
+        CarpetProfiler.start_section(world_name, "Villages");
         this.villageCollection.tick();
         this.villageSiege.tick();
+        CarpetProfiler.end_current_section();
         this.profiler.endStartSection("portalForcer");
         this.worldTeleporter.tick(this.getGameTime());
+        }
+        // [TISCM] Newlight
+        if (CarpetSettings.newLight)
+        {
+            this.profiler.endStartSection("lighting");
+            this.getLightingEngine().procLightUpdates();
+        }
         this.profiler.endSection();
         this.sendQueuedBlockEvents();
         this.insideTick = false;
@@ -270,7 +331,7 @@
                 }
             }
 
-            this.allPlayersSleeping = j > 0 && j >= this.playerEntities.size() - i;
+            this.allPlayersSleeping = j > 0 && (CarpetSettings.onePlayerSleeping || j >= this.playerEntities.size() - i);
         }
     }
 
@@ -306,6 +367,13 @@
     {
         if (this.allPlayersSleeping && !this.isRemote)
         {
+            if (CarpetSettings.onePlayerSleeping)
+            {
+                for (EntityPlayer entityplayer : this.playerEntities)
+                    if (!entityplayer.isSpectator() && entityplayer.isPlayerFullyAsleep())
+                        return true;
+                return false;
+            }
             for (EntityPlayer entityplayer : this.playerEntities)
             {
                 if (!entityplayer.isSpectator() && !entityplayer.isPlayerFullyAsleep())
@@ -371,7 +439,7 @@
             int j = MathHelper.floor(entityplayer.posX) + this.rand.nextInt(11) - 5;
             int k = MathHelper.floor(entityplayer.posY) + this.rand.nextInt(11) - 5;
             int l = MathHelper.floor(entityplayer.posZ) + this.rand.nextInt(11) - 5;
-            this.checkLight(new BlockPos(j, k, l));
+            this.checkLight(mutableBlockPosForRandomTicks.setPos(j, k, l));
         }
 
         this.profiler.endSection();
@@ -407,6 +475,13 @@
                 chunk.enqueueRelightChecks();
                 this.profiler.endStartSection("tickChunk");
                 chunk.tick(false);
+                //[CM] tick freeze
+                if (!TickSpeed.process_entities)
+                { // skipping the rest of the block processing
+                    this.profiler.endSection();
+                    continue;
+                }
+
                 this.profiler.endStartSection("thunder");
 
                 if (flag && flag1 && this.rand.nextInt(100000) == 0)
@@ -439,7 +514,9 @@
                 {
                     this.updateLCG = this.updateLCG * 3 + 1013904223;
                     int i2 = this.updateLCG >> 2;
-                    BlockPos blockpos1 = this.getHeight(Heightmap.Type.MOTION_BLOCKING, new BlockPos(j + (i2 & 15), 0, k + (i2 >> 8 & 15)));
+                    mutableBlockPosForRandomTicks.setPos(j + (i2 & 15), 0, k + (i2 >> 8 & 15));
+                    mutableBlockPosForRandomTicks.setY(this.getHeight(Heightmap.Type.MOTION_BLOCKING, mutableBlockPosForRandomTicks.getX(), mutableBlockPosForRandomTicks.getZ()));
+                    BlockPos blockpos1 = mutableBlockPosForRandomTicks;
                     BlockPos blockpos2 = blockpos1.down();
                     Biome biome = this.getBiome(blockpos1);
 
@@ -480,12 +557,12 @@
 
                                 if (iblockstate.ticksRandomly())
                                 {
-                                    iblockstate.randomTick(this, new BlockPos(j1 + j, l1 + chunksection.getYLocation(), k1 + k), this.rand);
+                                    iblockstate.randomTick(this, mutableBlockPosForRandomTicks.setPos(j1 + j, l1 + chunksection.getYLocation(), k1 + k), this.rand);
                                 }
 
                                 if (ifluidstate.ticksRandomly())
                                 {
-                                    ifluidstate.randomTick(this, new BlockPos(j1 + j, l1 + chunksection.getYLocation(), k1 + k), this.rand);
+                                    ifluidstate.randomTick(this, mutableBlockPosForRandomTicks.setPos(j1 + j, l1 + chunksection.getYLocation(), k1 + k), this.rand);
                                 }
 
                                 this.profiler.endSection();
@@ -499,7 +576,7 @@
         }
     }
 
-    protected BlockPos adjustPosToNearbyEntity(BlockPos pos)
+    public BlockPos adjustPosToNearbyEntity(BlockPos pos) // [CM] Changed access to public for summonNaturalLightning
     {
         BlockPos blockpos = this.getHeight(Heightmap.Type.MOTION_BLOCKING, pos);
         AxisAlignedBB axisalignedbb = (new AxisAlignedBB(blockpos, new BlockPos(blockpos.getX(), this.getHeight(), blockpos.getZ()))).grow(3.0D);
@@ -608,7 +685,14 @@
     {
         if (this.worldInfo.getGenerator() != WorldType.DEBUG_ALL_BLOCK_STATES)
         {
+            // TISCM TileTick logger
+            TileTickListLogHelper.setListInfo("Block", this.getDimension().getType().getId());
+            // end TISCM TileTick logger
             this.pendingBlockTicks.tick();
+
+            // TISCM TileTick logger
+            TileTickListLogHelper.setListInfo("Fluid", this.getDimension().getType().getId());
+            // end TISCM TileTick logger
             this.pendingFluidTicks.tick();
         }
     }
@@ -1032,7 +1116,7 @@
 
             if (this.fireBlockEvent(blockeventdata))
             {
-                this.server.getPlayerList().sendToAllNearExcept((EntityPlayer)null, (double)blockeventdata.getPosition().getX(), (double)blockeventdata.getPosition().getY(), (double)blockeventdata.getPosition().getZ(), 64.0D, this.dimension.getType(), new SPacketBlockAction(blockeventdata.getPosition(), blockeventdata.getBlock(), blockeventdata.getEventID(), blockeventdata.getEventParameter()));
+                this.server.getPlayerList().sendToAllNearExcept((EntityPlayer)null, (double)blockeventdata.getPosition().getX(), (double)blockeventdata.getPosition().getY(), (double)blockeventdata.getPosition().getZ(), CarpetSettings.blockEventPacketRange, this.dimension.getType(), new SPacketBlockAction(blockeventdata.getPosition(), blockeventdata.getBlock(), blockeventdata.getEventID(), blockeventdata.getEventParameter()));
             }
         }
     }
