--- a/net/minecraft/world/WorldServer.java
+++ b/net/minecraft/world/WorldServer.java
@@ -1,5 +1,13 @@
 package net.minecraft.world;
 
+import carpet.helpers.TickSpeed;
+import carpet.microtick.MicroTickLogger;
+import carpet.microtick.MicroTickLoggerManager;
+import carpet.microtick.enums.EventType;
+import carpet.microtick.enums.TickStage;
+import carpet.microtick.tickstages.BlockEventTickStageExtra;
+import carpet.settings.CarpetSettings;
+import carpet.utils.CarpetProfiler;
 import com.google.common.collect.Lists;
 import com.google.common.collect.Maps;
 import com.google.common.util.concurrent.ListenableFuture;
@@ -14,6 +22,8 @@
 import java.util.stream.Stream;
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
+
+import me.jellysquid.mods.lithium.common.world.scheduler.LithiumServerTickScheduler;
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockEventData;
 import net.minecraft.block.state.IBlockState;
@@ -93,19 +103,23 @@
     private final EntityTracker entityTracker;
     private final PlayerChunkMap playerChunkMap;
     private final Map<UUID, Entity> entitiesByUuid = Maps.newHashMap();
+    private final MicroTickLogger microTickLogger;  // TISCM Micro Tick logger
     public boolean disableLevelSaving;
     private boolean allPlayersSleeping;
     private int updateEntityTick;
     private final Teleporter worldTeleporter;
     private final WorldEntitySpawner entitySpawner = new WorldEntitySpawner();
-    private final ServerTickList<Block> pendingBlockTicks = new ServerTickList<>(this, (p_205341_0_) ->
+
+    // Lithium world.tick_scheduler
+    private final ServerTickList<Block> pendingBlockTicks = new LithiumServerTickScheduler<>(this, (p_205341_0_) ->
     {
         return p_205341_0_ == null || p_205341_0_.getDefaultState().isAir();
     }, IRegistry.BLOCK::getKey, IRegistry.BLOCK::getOrDefault, this::tickBlock);
-    private final ServerTickList<Fluid> pendingFluidTicks = new ServerTickList<>(this, (p_205774_0_) ->
+    private final ServerTickList<Fluid> pendingFluidTicks = new LithiumServerTickScheduler<>(this, (p_205774_0_) ->
     {
         return p_205774_0_ == null || p_205774_0_ == Fluids.EMPTY;
     }, IRegistry.FLUID::getKey, IRegistry.FLUID::getOrDefault, this::tickFluid);
+
     protected final VillageSiege villageSiege = new VillageSiege(this);
     ObjectLinkedOpenHashSet<BlockEventData> blockEventQueue = new ObjectLinkedOpenHashSet<>();
     private boolean insideTick;
@@ -122,8 +136,15 @@
         this.calculateInitialSkylight();
         this.calculateInitialWeather();
         this.getWorldBorder().setSize(server.getMaxWorldSize());
+
+        this.microTickLogger = new MicroTickLogger(this);  // TISCM Micro Tick logger
     }
 
+    public MicroTickLogger getMicroTickLogger()  // TISCM Micro Tick logger
+    {
+        return microTickLogger;
+    }
+
     public WorldServer init()
     {
         String s = VillageCollection.fileNameForProvider(this.dimension);
@@ -191,15 +212,28 @@
             this.wakeAllPlayers();
         }
 
+        //CM profiler
+        int did = this.dimension.getType().getId();
+        String world_name = (did==0)?"Overworld":((did<0?"The Nether":"The End"));
+        // CM end
+
+        if (TickSpeed.process_entities)
+        { // [CM] extra indent to skip processing of entities
         this.profiler.startSection("spawner");
+        CarpetProfiler.start_section(world_name, "Spawning");
+        MicroTickLoggerManager.setTickStage(this, TickStage.SPAWNING); // TISCM Micro Tick logger
 
         if (this.getGameRules().getBoolean("doMobSpawning") && this.worldInfo.getGenerator() != WorldType.DEBUG_ALL_BLOCK_STATES)
         {
             this.entitySpawner.findChunksForSpawning(this, this.spawnHostileMobs, this.spawnPeacefulMobs, this.worldInfo.getGameTime() % 400L == 0L);
             this.getChunkProvider().spawnMobs(this, this.spawnHostileMobs, this.spawnPeacefulMobs);
         }
+        CarpetProfiler.end_current_section();
+        }
+        // [CM] end extra indent
 
         this.profiler.endStartSection("chunkSource");
+        MicroTickLoggerManager.setTickStage(this, TickStage.CHUNK_UNLOADING); // TISCM Micro Tick logger
         this.chunkProvider.tick(hasTimeLeft);
         int j = this.calculateSkylightSubtracted(1.0F);
 
@@ -208,25 +242,57 @@
             this.setSkylightSubtracted(j);
         }
 
-        this.worldInfo.setGameTime(this.worldInfo.getGameTime() + 1L);
+        if (TickSpeed.process_entities)
+        { // CM extra indent to skip processing of entities
 
+        // TISCM Micro Tick logger
+        // Flushes messages right before overworld gametime increases
+        if (this.getDimension().getType() == DimensionType.OVERWORLD)
+        {
+            MicroTickLoggerManager.flushMessages();
+        }
+
+        this.worldInfo.setGameTime(this.worldInfo.getGameTime() + 1L);
         if (this.getGameRules().getBoolean("doDaylightCycle"))
         {
             this.worldInfo.setDayTime(this.worldInfo.getDayTime() + 1L);
         }
 
         this.profiler.endStartSection("tickPending");
+        CarpetProfiler.start_section(world_name, "Blocks");
+        MicroTickLoggerManager.setTickStage(this, TickStage.TILE_TICK); // TISCM Micro Tick logger
         this.tickPending();
+        CarpetProfiler.end_current_section();
+        } // end extra indent
         this.profiler.endStartSection("tickBlocks");
+        CarpetProfiler.start_section(world_name, "Blocks");
         this.tickBlocks();
+        CarpetProfiler.end_current_section();
+        MicroTickLoggerManager.setTickStage(this, TickStage.PLAYER_CHUNK_MAP); // TISCM Micro Tick logger
         this.profiler.endStartSection("chunkMap");
         this.playerChunkMap.tick();
+
+        if (TickSpeed.process_entities)
+        { // CM extra indent to skip processing of entities
         this.profiler.endStartSection("village");
+        CarpetProfiler.start_section(world_name, "Villages");
+        MicroTickLoggerManager.setTickStage(this, TickStage.VILLAGE); // TISCM Micro Tick logger
         this.villageCollection.tick();
         this.villageSiege.tick();
+        CarpetProfiler.end_current_section();
         this.profiler.endStartSection("portalForcer");
+        MicroTickLoggerManager.setTickStage(this, TickStage.PORTAL_FORCER); // TISCM Micro Tick logger
         this.worldTeleporter.tick(this.getGameTime());
+        }
+        // [TISCM] Newlight
+        if (CarpetSettings.newLight)
+        {
+            this.profiler.endStartSection("lighting");
+            MicroTickLoggerManager.setTickStage(this, TickStage.NEW_LIGHT); // TISCM Micro Tick logger
+            this.getLightingEngine().procLightUpdates();
+        }
         this.profiler.endSection();
+        MicroTickLoggerManager.setTickStage(this, TickStage.BLOCK_EVENT); // TISCM Micro Tick logger
         this.sendQueuedBlockEvents();
         this.insideTick = false;
     }
@@ -270,7 +336,7 @@
                 }
             }
 
-            this.allPlayersSleeping = j > 0 && j >= this.playerEntities.size() - i;
+            this.allPlayersSleeping = j > 0 && (CarpetSettings.onePlayerSleeping || j >= this.playerEntities.size() - i);
         }
     }
 
@@ -306,6 +372,13 @@
     {
         if (this.allPlayersSleeping && !this.isRemote)
         {
+            if (CarpetSettings.onePlayerSleeping)
+            {
+                for (EntityPlayer entityplayer : this.playerEntities)
+                    if (!entityplayer.isSpectator() && entityplayer.isPlayerFullyAsleep())
+                        return true;
+                return false;
+            }
             for (EntityPlayer entityplayer : this.playerEntities)
             {
                 if (!entityplayer.isSpectator() && !entityplayer.isPlayerFullyAsleep())
@@ -397,9 +470,11 @@
             boolean flag1 = this.isThundering();
             this.profiler.startSection("pollingChunks");
 
+            MicroTickLoggerManager.setTickStage(this, TickStage.RANDOMTICK_CLIMATE); // TISCM Micro Tick logger
             for (Iterator<Chunk> iterator = this.playerChunkMap.getChunkIterator(); iterator.hasNext(); this.profiler.endSection())
             {
                 this.profiler.startSection("getChunk");
+                MicroTickLoggerManager.setTickStageDetail(this, "TickChunk"); // TISCM Micro Tick logger
                 Chunk chunk = iterator.next();
                 int j = chunk.x * 16;
                 int k = chunk.z * 16;
@@ -407,7 +482,15 @@
                 chunk.enqueueRelightChecks();
                 this.profiler.endStartSection("tickChunk");
                 chunk.tick(false);
+                //[CM] tick freeze
+                if (!TickSpeed.process_entities)
+                { // skipping the rest of the block processing
+                    this.profiler.endSection();
+                    continue;
+                }
+
                 this.profiler.endStartSection("thunder");
+                MicroTickLoggerManager.setTickStageDetail(this, "Thunder"); // TISCM Micro Tick logger
 
                 if (flag && flag1 && this.rand.nextInt(100000) == 0)
                 {
@@ -434,6 +517,7 @@
                 }
 
                 this.profiler.endStartSection("iceandsnow");
+                MicroTickLoggerManager.setTickStageDetail(this, "Ice&Snow"); // TISCM Micro Tick logger
 
                 if (this.rand.nextInt(16) == 0)
                 {
@@ -460,6 +544,7 @@
                 }
 
                 this.profiler.endStartSection("tickBlocks");
+                MicroTickLoggerManager.setTickStageDetail(this, "RandomTick"); // TISCM Micro Tick logger
 
                 if (i > 0)
                 {
@@ -496,10 +581,15 @@
             }
 
             this.profiler.endSection();
+
+            // TISCM Micro Tick logger
+            MicroTickLoggerManager.setTickStageDetail(this, null);
+            MicroTickLoggerManager.setTickStageExtra(this, null);
+            // end TISCM Micro Tick logger
         }
     }
 
-    protected BlockPos adjustPosToNearbyEntity(BlockPos pos)
+    public BlockPos adjustPosToNearbyEntity(BlockPos pos) // [CM] Changed access to public for summonNaturalLightning
     {
         BlockPos blockpos = this.getHeight(Heightmap.Type.MOTION_BLOCKING, pos);
         AxisAlignedBB axisalignedbb = (new AxisAlignedBB(blockpos, new BlockPos(blockpos.getX(), this.getHeight(), blockpos.getZ()))).grow(3.0D);
@@ -625,12 +715,16 @@
 
     private void tickBlock(NextTickListEntry<Block> blockTickEntry)
     {
+        MicroTickLoggerManager.onExecuteTileTickEvent(this, blockTickEntry, EventType.ACTION_START);  // TISCM Micro Tick logger
+
         IBlockState iblockstate = this.getBlockState(blockTickEntry.position);
 
         if (iblockstate.getBlock() == blockTickEntry.getTarget())
         {
             iblockstate.tick(this, blockTickEntry.position, this.rand);
         }
+
+        MicroTickLoggerManager.onExecuteTileTickEvent(this, blockTickEntry, EventType.ACTION_END);  // TISCM Micro Tick logger
     }
 
     public void tickEntity(Entity entityIn, boolean forceUpdate)
@@ -1021,26 +1115,54 @@
 
     public void addBlockEvent(BlockPos pos, Block blockIn, int eventID, int eventParam)
     {
-        this.blockEventQueue.add(new BlockEventData(pos, blockIn, eventID, eventParam));
+        // TISCM Micro Tick logger
+        BlockEventData blockEventData = new BlockEventData(pos, blockIn, eventID, eventParam);
+        MicroTickLoggerManager.onScheduleBlockEvent(this, blockEventData);
+        this.blockEventQueue.add(blockEventData);
     }
 
     private void sendQueuedBlockEvents()
     {
+        // TISCM Micro Tick logger
+        int depth = 0, depthSize = this.blockEventQueue.size();
+        int eventCounter = 0, indexInCurrentDepth = 0;
+        // end TISCM Micro Tick logger
+
         while (!this.blockEventQueue.isEmpty())
         {
             BlockEventData blockeventdata = this.blockEventQueue.removeFirst();
 
+            // TISCM Micro Tick logger
+            MicroTickLoggerManager.setTickStageDetail(this, String.valueOf(depth));
+            MicroTickLoggerManager.setTickStageExtra(this, new BlockEventTickStageExtra(this, blockeventdata, eventCounter++, depth));
+            // end TISCM Micro Tick logger
+
             if (this.fireBlockEvent(blockeventdata))
             {
-                this.server.getPlayerList().sendToAllNearExcept((EntityPlayer)null, (double)blockeventdata.getPosition().getX(), (double)blockeventdata.getPosition().getY(), (double)blockeventdata.getPosition().getZ(), 64.0D, this.dimension.getType(), new SPacketBlockAction(blockeventdata.getPosition(), blockeventdata.getBlock(), blockeventdata.getEventID(), blockeventdata.getEventParameter()));
+                this.server.getPlayerList().sendToAllNearExcept((EntityPlayer)null, (double)blockeventdata.getPosition().getX(), (double)blockeventdata.getPosition().getY(), (double)blockeventdata.getPosition().getZ(), CarpetSettings.blockEventPacketRange, this.dimension.getType(), new SPacketBlockAction(blockeventdata.getPosition(), blockeventdata.getBlock(), blockeventdata.getEventID(), blockeventdata.getEventParameter()));
             }
+
+            // TISCM Micro Tick logger
+            indexInCurrentDepth++;
+            if (indexInCurrentDepth == depthSize)
+            {
+                depth++;
+                depthSize = this.blockEventQueue.size();
+                indexInCurrentDepth = 0;
+            }
+            // end TISCM Micro Tick logger
         }
     }
 
     private boolean fireBlockEvent(BlockEventData event)
     {
+        MicroTickLoggerManager.onExecuteBlockEvent(this, event, null, EventType.ACTION_START);  // TISCM Micro Tick logger
+
         IBlockState iblockstate = this.getBlockState(event.getPosition());
-        return iblockstate.getBlock() == event.getBlock() ? iblockstate.onBlockEventReceived(this, event.getPosition(), event.getEventID(), event.getEventParameter()) : false;
+        boolean returnValue = iblockstate.getBlock() == event.getBlock() ? iblockstate.onBlockEventReceived(this, event.getPosition(), event.getEventID(), event.getEventParameter()) : false;
+
+        MicroTickLoggerManager.onExecuteBlockEvent(this, event, returnValue, EventType.ACTION_END);  // TISCM Micro Tick logger
+        return returnValue;
     }
 
     public void close()
