--- a/net/minecraft/util/math/shapes/DoubleRangeList.java
+++ b/net/minecraft/util/math/shapes/DoubleRangeList.java
@@ -1,19 +1,32 @@
 package net.minecraft.util.math.shapes;
 
 import it.unimi.dsi.fastutil.doubles.AbstractDoubleList;
+import me.jellysquid.mods.lithium.LithiumConfig;
 
-class DoubleRangeList extends AbstractDoubleList
+// it's FractionalDoubleList in yarn mapping
+// Lithium shapes.specialized_shapes makes public
+public class DoubleRangeList extends AbstractDoubleList
 {
     private final int field_197854_a;
+    // Lithium shapes.precompute_shape_arrays
+    private final double scale;
 
-    DoubleRangeList(int p_i47689_1_)
+    // Lithium shapes.specialized_shapes makes public
+    public DoubleRangeList(int p_i47689_1_)
     {
         this.field_197854_a = p_i47689_1_;
+
+        // Lithium shapes.precompute_shape_arrays
+        if (LithiumConfig.SHAPES_PRECOMPUTE_SHAPE_ARRAYS)
+        {
+            this.scale = 1.0D / this.field_197854_a;
+        }
     }
 
     public double getDouble(int p_getDouble_1_)
     {
-        return (double)p_getDouble_1_ / (double)this.field_197854_a;
+        // Lithium shapes.precompute_shape_arrays
+        return LithiumConfig.SHAPES_PRECOMPUTE_SHAPE_ARRAYS ? p_getDouble_1_ * this.scale : (double)p_getDouble_1_ / (double)this.field_197854_a;
     }
 
     public int size()
