--- a/net/minecraft/util/math/shapes/VoxelShapeInt.java
+++ b/net/minecraft/util/math/shapes/VoxelShapeInt.java
@@ -1,6 +1,8 @@
 package net.minecraft.util.math.shapes;
 
+import carpet.settings.CarpetSettings;
 import it.unimi.dsi.fastutil.doubles.DoubleList;
+import me.jellysquid.mods.lithium.LithiumConfig;
 import net.minecraft.util.EnumFacing;
 
 public final class VoxelShapeInt extends VoxelShape
@@ -9,16 +11,45 @@
     private final int y;
     private final int z;
 
+    // Lithium shapes.precompute_shape_arrays
+    private static final EnumFacing.Axis[] AXIS = EnumFacing.Axis.values();
+
+    @SuppressWarnings("FieldMayBeFinal")
+    private DoubleList[] list;
+
+
     public VoxelShapeInt(VoxelShapePart p_i47679_1_, int x, int y, int z)
     {
         super(p_i47679_1_);
         this.x = x;
         this.y = y;
         this.z = z;
+
+        // Lithium shapes.precompute_shape_arrays
+        if (LithiumConfig.SHAPES_PRECOMPUTE_SHAPE_ARRAYS)
+        {
+            this.list = new DoubleList[AXIS.length];
+
+            for (EnumFacing.Axis axis : AXIS)
+            {
+                this.list[axis.ordinal()] = new IntRangeList(this.part.getSize(axis), axis.getCoordinate(this.x, this.y, this.z));
+            }
+        }
     }
 
     protected DoubleList getValues(EnumFacing.Axis axis)
     {
-        return new IntRangeList(this.part.getSize(axis), axis.getCoordinate(this.x, this.y, this.z));
+        // Lithium shapes.precompute_shape_arrays
+        return LithiumConfig.SHAPES_PRECOMPUTE_SHAPE_ARRAYS ? this.list[axis.ordinal()] : new IntRangeList(this.part.getSize(axis), axis.getCoordinate(this.x, this.y, this.z));
     }
+    
+    // [CM] start
+    @Override
+    public int getClosestIndex(final EnumFacing.Axis axis, final double position)
+    {
+        return !CarpetSettings.optimizeVoxelCode ?
+                       super.getClosestIndex(axis, position) :
+                       net.minecraft.util.math.MathHelper.clamp(net.minecraft.util.math.MathHelper.floor(position - axis.getCoordinate(this.x, this.y, this.z)), -1, this.part.getSize(axis));
+    }
+    // [CM] end
 }
