--- a/net/minecraft/util/math/shapes/VoxelShapeCube.java
+++ b/net/minecraft/util/math/shapes/VoxelShapeCube.java
@@ -1,9 +1,11 @@
 package net.minecraft.util.math.shapes;
 
+import carpet.settings.CarpetSettings;
 import it.unimi.dsi.fastutil.doubles.DoubleList;
 import net.minecraft.util.EnumFacing;
 
-final class VoxelShapeCube extends VoxelShape
+// Lithium shapes.specialized_shapes makes public
+public final class VoxelShapeCube extends VoxelShape
 {
     VoxelShapeCube(VoxelShapePart p_i48182_1_)
     {
@@ -14,4 +16,14 @@
     {
         return new DoubleRangeList(this.part.getSize(axis));
     }
+    
+    // [CM] start
+    @Override
+    public int getClosestIndex(final EnumFacing.Axis axis, final double position)
+    {
+        return !CarpetSettings.optimizeVoxelCode ?
+                       super.getClosestIndex(axis, position) :
+                       net.minecraft.util.math.MathHelper.clamp(net.minecraft.util.math.MathHelper.floor(position * this.part.getSize(axis)), -1, this.part.getSize(axis));
+    }
+    // [CM] end
 }
