--- a/net/minecraft/fluid/LavaFluid.java
+++ b/net/minecraft/fluid/LavaFluid.java
@@ -127,7 +127,7 @@
 
     private boolean isSurroundingBlockFlammable(IWorldReaderBase worldIn, BlockPos pos)
     {
-        for (EnumFacing enumfacing : EnumFacing.values())
+        for (EnumFacing enumfacing : (me.jellysquid.mods.lithium.LithiumConfig.ALLOC_ENUM_VALUES ? EnumFacing.ENUM_FACING_VALUES : EnumFacing.values()))
         {
             if (this.getCanBlockBurn(worldIn, pos.offset(enumfacing)))
             {
