--- a/net/minecraft/entity/boss/EntityWither.java
+++ b/net/minecraft/entity/boss/EntityWither.java
@@ -1,5 +1,6 @@
 package net.minecraft.entity.boss;
 
+import carpet.settings.CarpetSettings;
 import com.google.common.collect.ImmutableList;
 import java.util.List;
 import java.util.function.Predicate;
@@ -246,7 +247,12 @@
             if (j1 <= 0)
             {
                 this.world.newExplosion(this, this.posX, this.posY + (double)this.getEyeHeight(), this.posZ, 7.0F, false, this.world.getGameRules().getBoolean("mobGriefing"));
-                this.world.playBroadcastSound(1023, new BlockPos(this), 0);
+
+                // TISCM witherSpawnedSoundDisabled wraps the invocation
+                if (!CarpetSettings.witherSpawnedSoundDisabled)
+                {
+                    this.world.playBroadcastSound(1023, new BlockPos(this), 0);
+                }
             }
 
             this.setInvulTime(j1);
