--- a/net/minecraft/entity/monster/EntityShulker.java
+++ b/net/minecraft/entity/monster/EntityShulker.java
@@ -247,7 +247,7 @@
             {
                 boolean flag = false;
 
-                for (EnumFacing enumfacing1 : EnumFacing.values())
+                for (EnumFacing enumfacing1 : (me.jellysquid.mods.lithium.LithiumConfig.ALLOC_ENUM_VALUES ? EnumFacing.ENUM_FACING_VALUES : EnumFacing.values()))
                 {
                     blockpos1 = blockpos.offset(enumfacing1);
 
@@ -405,7 +405,7 @@
                 {
                     boolean flag = false;
 
-                    for (EnumFacing enumfacing : EnumFacing.values())
+                    for (EnumFacing enumfacing : (me.jellysquid.mods.lithium.LithiumConfig.ALLOC_ENUM_VALUES ? EnumFacing.ENUM_FACING_VALUES : EnumFacing.values()))
                     {
                         if (this.world.isTopSolid(blockpos1.offset(enumfacing)))
                         {
