--- a/net/minecraft/entity/monster/EntityMob.java
+++ b/net/minecraft/entity/monster/EntityMob.java
@@ -1,5 +1,6 @@
 package net.minecraft.entity.monster;
 
+import carpet.commands.lifetime.removal.LiteralRemovalReason;
 import net.minecraft.entity.EntityCreature;
 import net.minecraft.entity.EntityType;
 import net.minecraft.entity.SharedMonsterAttributes;
@@ -47,6 +48,9 @@
 
         if (!this.world.isRemote && this.world.getDifficulty() == EnumDifficulty.PEACEFUL)
         {
+            // TISCM lifetime tracker
+            this.recordRemoval(LiteralRemovalReason.DESPAWN_DIFFICULTY);
+
             this.remove();
         }
     }
