--- a/net/minecraft/block/BlockSponge.java
+++ b/net/minecraft/block/BlockSponge.java
@@ -55,7 +55,7 @@
             BlockPos blockpos = tuple.getA();
             int j = tuple.getB();
 
-            for (EnumFacing enumfacing : EnumFacing.values())
+            for (EnumFacing enumfacing : (me.jellysquid.mods.lithium.LithiumConfig.ALLOC_ENUM_VALUES ? EnumFacing.ENUM_FACING_VALUES : EnumFacing.values()))
             {
                 BlockPos blockpos1 = blockpos.offset(enumfacing);
                 IBlockState iblockstate = worldIn.getBlockState(blockpos1);
