--- a/net/minecraft/block/BlockShulkerBox.java
+++ b/net/minecraft/block/BlockShulkerBox.java
@@ -2,6 +2,8 @@
 
 import java.util.List;
 import javax.annotation.Nullable;
+
+import carpet.settings.CarpetSettings;
 import net.minecraft.block.material.EnumPushReaction;
 import net.minecraft.block.state.BlockFaceShape;
 import net.minecraft.block.state.IBlockState;
@@ -104,6 +106,12 @@
                 {
                     AxisAlignedBB axisalignedbb = VoxelShapes.fullCube().getBoundingBox().expand((double)(0.5F * (float)enumfacing.getXOffset()), (double)(0.5F * (float)enumfacing.getYOffset()), (double)(0.5F * (float)enumfacing.getZOffset())).contract((double)enumfacing.getXOffset(), (double)enumfacing.getYOffset(), (double)enumfacing.getZOffset());
                     flag = worldIn.isCollisionBoxesEmpty((Entity)null, axisalignedbb.offset(pos.offset(enumfacing)));
+
+                    // TISCM creativeOpenContainerForcibly
+                    if (CarpetSettings.creativeOpenContainerForcibly)
+                    {
+                        flag |= player.isCreative();
+                    }
                 }
                 else
                 {
