--- a/net/minecraft/block/BlockCoralPlantBase.java
+++ b/net/minecraft/block/BlockCoralPlantBase.java
@@ -48,7 +48,7 @@
         }
         else
         {
-            for (EnumFacing enumfacing : EnumFacing.values())
+            for (EnumFacing enumfacing : (me.jellysquid.mods.lithium.LithiumConfig.ALLOC_ENUM_VALUES ? EnumFacing.ENUM_FACING_VALUES : EnumFacing.values()))
             {
                 if (worldIn.getFluidState(pos.offset(enumfacing)).isTagged(FluidTags.WATER))
                 {
