--- a/net/minecraft/block/BlockBasePressurePlate.java
+++ b/net/minecraft/block/BlockBasePressurePlate.java
@@ -1,6 +1,9 @@
 package net.minecraft.block;
 
 import java.util.Random;
+
+import carpet.microtiming.MicroTimingLoggerManager;
+import carpet.microtiming.enums.EventType;
 import net.minecraft.block.material.EnumPushReaction;
 import net.minecraft.block.state.BlockFaceShape;
 import net.minecraft.block.state.IBlockState;
@@ -131,8 +134,10 @@
 
     protected void updateNeighbors(World worldIn, BlockPos pos)
     {
+        MicroTimingLoggerManager.onEmitBlockUpdate(worldIn, this, pos, EventType.ACTION_START, "updateNeighbors");  // TISCM Micro Timing logger
         worldIn.notifyNeighborsOfStateChange(pos, this);
         worldIn.notifyNeighborsOfStateChange(pos.down(), this);
+        MicroTimingLoggerManager.onEmitBlockUpdate(worldIn, this, pos, EventType.ACTION_END, "updateNeighbors");  // TISCM Micro Timing logger
     }
 
     public int getWeakPower(IBlockState blockState, IBlockReader blockAccess, BlockPos pos, EnumFacing side)
