--- a/net/minecraft/block/BlockPressurePlateWeighted.java
+++ b/net/minecraft/block/BlockPressurePlateWeighted.java
@@ -13,8 +13,10 @@
 import net.minecraft.world.IWorld;
 import net.minecraft.world.IWorldReaderBase;
 import net.minecraft.world.World;
+import redstone.multimeter.block.Meterable;
+import redstone.multimeter.block.PowerSource;
 
-public class BlockPressurePlateWeighted extends BlockBasePressurePlate
+public class BlockPressurePlateWeighted extends BlockBasePressurePlate implements Meterable /*RSMM*/, PowerSource /*RSMM*/
 {
     public static final IntegerProperty POWER = BlockStateProperties.POWER_0_15;
     private final int maxWeight;
@@ -70,4 +72,16 @@
     {
         builder.add(POWER);
     }
+
+    // RSMM
+    @Override
+    public boolean isActive(World world, BlockPos pos, IBlockState state) {
+        return state.get(POWER) > MIN_POWER;
+    }
+
+    // RSMM
+    @Override
+    public int getPowerLevel(World world, BlockPos pos, IBlockState state) {
+        return state.get(POWER);
+    }
 }
