--- a/net/minecraft/block/BlockCommandBlock.java
+++ b/net/minecraft/block/BlockCommandBlock.java
@@ -1,6 +1,8 @@
 package net.minecraft.block;
 
 import java.util.Random;
+
+import carpet.settings.CarpetSettings;
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.entity.EntityLivingBase;
 import net.minecraft.entity.player.EntityPlayer;
@@ -21,10 +23,7 @@
 import net.minecraft.util.Rotation;
 import net.minecraft.util.StringUtils;
 import net.minecraft.util.math.BlockPos;
-import net.minecraft.world.GameRules;
-import net.minecraft.world.IBlockReader;
-import net.minecraft.world.IWorldReaderBase;
-import net.minecraft.world.World;
+import net.minecraft.world.*;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
@@ -65,7 +64,18 @@
                     if (flag)
                     {
                         tileentitycommandblock.setConditionMet();
-                        worldIn.getPendingBlockTicks().scheduleTick(pos, this, this.tickRate(worldIn));
+                        if (CarpetSettings.instantCommandBlock && tileentitycommandblock.getMode() == TileEntityCommandBlock.Mode.REDSTONE)  // TISCM instantCommandBlock
+                        {
+                            Block blockBelow = worldIn.getBlockState(pos.down()).getBlock();
+                            if (blockBelow == Blocks.REDSTONE_ORE)
+                            {
+                                this.tick(state, worldIn, pos, worldIn.getRandom());
+                            }
+                        }
+                        else
+                        {
+                            worldIn.getPendingBlockTicks().scheduleTick(pos, this, this.tickRate(worldIn));
+                        }
                     }
                 }
             }
