--- a/net/minecraft/block/BlockObserver.java
+++ b/net/minecraft/block/BlockObserver.java
@@ -1,6 +1,9 @@
 package net.minecraft.block;
 
 import java.util.Random;
+
+import carpet.microtick.MicroTickLoggerManager;
+import carpet.microtick.enums.EventType;
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.item.BlockItemUseContext;
 import net.minecraft.state.BooleanProperty;
@@ -74,10 +77,12 @@
 
     protected void updateNeighborsInFront(World worldIn, BlockPos pos, IBlockState state)
     {
+        MicroTickLoggerManager.onEmitBlockUpdate(worldIn, this, pos, EventType.ACTION_START, "updateNeighborsInFront");  // TISCM Micro Tick logger
         EnumFacing enumfacing = state.get(FACING);
         BlockPos blockpos = pos.offset(enumfacing.getOpposite());
         worldIn.neighborChanged(blockpos, this, pos);
         worldIn.notifyNeighborsOfStateExcept(blockpos, this, enumfacing);
+        MicroTickLoggerManager.onEmitBlockUpdate(worldIn, this, pos, EventType.ACTION_END, "updateNeighborsInFront");  // TISCM Micro Tick logger
     }
 
     public boolean canProvidePower(IBlockState state)
