--- a/net/minecraft/block/BlockButton.java
+++ b/net/minecraft/block/BlockButton.java
@@ -3,6 +3,9 @@
 import java.util.List;
 import java.util.Random;
 import javax.annotation.Nullable;
+
+import carpet.microtiming.MicroTimingLoggerManager;
+import carpet.microtiming.enums.EventType;
 import net.minecraft.block.state.BlockFaceShape;
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.entity.Entity;
@@ -202,8 +205,10 @@
 
     private void updateNeighbors(IBlockState p_196368_1_, World p_196368_2_, BlockPos p_196368_3_)
     {
+        MicroTimingLoggerManager.onEmitBlockUpdate(p_196368_2_, this, p_196368_3_, EventType.ACTION_START, "updateNeighbors");  // TISCM Micro Timing logger
         p_196368_2_.notifyNeighborsOfStateChange(p_196368_3_, this);
         p_196368_2_.notifyNeighborsOfStateChange(p_196368_3_.offset(getFacing(p_196368_1_).getOpposite()), this);
+        MicroTimingLoggerManager.onEmitBlockUpdate(p_196368_2_, this, p_196368_3_, EventType.ACTION_END, "updateNeighbors");  // TISCM Micro Timing logger
     }
 
     protected void fillStateContainer(StateContainer.Builder<Block, IBlockState> builder)
