--- a/net/minecraft/block/BlockFlowingFluid.java
+++ b/net/minecraft/block/BlockFlowingFluid.java
@@ -152,7 +152,7 @@
         {
             boolean flag = false;
 
-            for (EnumFacing enumfacing : EnumFacing.values())
+            for (EnumFacing enumfacing : (me.jellysquid.mods.lithium.LithiumConfig.ALLOC_ENUM_VALUES ? EnumFacing.ENUM_FACING_VALUES : EnumFacing.values()))
             {
                 if (enumfacing != EnumFacing.DOWN && worldIn.getFluidState(pos.offset(enumfacing)).isTagged(FluidTags.WATER))
                 {
