--- a/net/minecraft/util/Util.java
+++ b/net/minecraft/util/Util.java
@@ -1,5 +1,7 @@
 package net.minecraft.util;
 
+import carpet.settings.CarpetSettings;
+import carpet.utils.deobfuscator.StackTraceDeobfuscator;
 import com.google.common.collect.Iterators;
 import it.unimi.dsi.fastutil.Hash.Strategy;
 import java.io.File;
@@ -17,10 +19,7 @@
 import java.security.PrivilegedActionException;
 import java.security.PrivilegedExceptionAction;
 import java.time.Instant;
-import java.util.Iterator;
-import java.util.List;
-import java.util.Locale;
-import java.util.Map;
+import java.util.*;
 import java.util.Map.Entry;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.FutureTask;
@@ -158,6 +157,10 @@
         }
         catch (ExecutionException executionexception)
         {
+            if (CarpetSettings.deobfuscateCrashReportStackTrace)
+            {
+                StackTraceDeobfuscator.deobfuscateThrowableStackTrace(executionexception);
+            }
             logger.fatal("Error executing task", (Throwable)executionexception);
         }
         catch (InterruptedException interruptedexception)
