--- a/net/minecraft/network/PacketThreadUtil.java
+++ b/net/minecraft/network/PacketThreadUtil.java
@@ -1,5 +1,8 @@
 package net.minecraft.network;
 
+import carpet.logging.microtiming.MicroTimingLoggerManager;
+import carpet.logging.microtiming.enums.TickStage;
+import carpet.logging.microtiming.tickphase.substages.PlayerActionSubStage;
 import net.minecraft.util.IThreadListener;
 
 public class PacketThreadUtil
@@ -10,7 +13,18 @@
         {
             scheduler.addScheduledTask(() ->
             {
+                // TISCM Micro Timing logger
+                if (processor instanceof NetHandlerPlayServer)
+                {
+                    NetHandlerPlayServer handler = (NetHandlerPlayServer)processor;
+                    MicroTimingLoggerManager.setTickStage(TickStage.PLAYER_ACTION);
+                    MicroTimingLoggerManager.setSubTickStage(new PlayerActionSubStage(handler.player));
+                }
+
                 packetIn.processPacket(processor);
+
+                MicroTimingLoggerManager.setSubTickStage(null);
+                MicroTimingLoggerManager.setTickStage(TickStage.ASYNC_TASK);
             });
             throw ThreadQuickExitException.INSTANCE;
         }
