--- a/net/minecraft/item/ItemBoat.java
+++ b/net/minecraft/item/ItemBoat.java
@@ -1,6 +1,8 @@
 package net.minecraft.item;
 
 import java.util.List;
+
+import carpet.settings.CarpetSettings;
 import net.minecraft.block.Block;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.item.EntityBoat;
@@ -72,6 +74,12 @@
                 }
             }
 
+            // TISCM entityPlacementIgnoreCollision removes entity collision check
+            if (CarpetSettings.entityPlacementIgnoreCollision)
+            {
+                flag = false;
+            }
+
             if (flag)
             {
                 return new ActionResult<>(EnumActionResult.PASS, itemstack);
@@ -84,7 +92,8 @@
                 entityboat.setBoatType(this.type);
                 entityboat.rotationYaw = playerIn.rotationYaw;
 
-                if (!worldIn.isCollisionBoxesEmpty(entityboat, entityboat.getBoundingBox().grow(-0.1D)))
+                // TISCM entityPlacementIgnoreCollision removes collision check
+                if (!worldIn.isCollisionBoxesEmpty(entityboat, entityboat.getBoundingBox().grow(-0.1D)) && !CarpetSettings.entityPlacementIgnoreCollision)
                 {
                     return new ActionResult<>(EnumActionResult.FAIL, itemstack);
                 }
