--- a/net/minecraft/world/gen/feature/GlowstoneFeature.java
+++ b/net/minecraft/world/gen/feature/GlowstoneFeature.java
@@ -32,7 +32,7 @@
                 {
                     int j = 0;
 
-                    for (EnumFacing enumfacing : EnumFacing.values())
+                    for (EnumFacing enumfacing : (me.jellysquid.mods.lithium.LithiumConfig.ALLOC_ENUM_VALUES ? EnumFacing.ENUM_FACING_VALUES : EnumFacing.values()))
                     {
                         if (worldIn.getBlockState(blockpos.offset(enumfacing)).getBlock() == Blocks.GLOWSTONE)
                         {
